### DISCUSSION TITLE
Regex lookbehind implementation decission

### DISCUSSION THREAD

**Discussion by MaxSagebaum** [hsutter/cppfront#1330] [Q&A]
Hello @ all,

I am currently implementing the missing regex lookbehind functionality. In this context I need to transform the iterators into reverse iterators, and the match context into a reverse match context.

The problem I have is the following: Currently, the function generation is strongly typed for the iterator (`Iter`) and the context.  I have now two ways to go:
 1) Extend the functionality of the code generation to a `reverse` generation. It would be some additional ifs that prepend a `Reverse` in front of the iterator type and the context type. This would increase code complexity for the generator and add kind of a statement machine. (Reverse generation on/off).
 2) I could remove the strong typing and use template type deduction for the iterator and context. This would simplify the generated code.

My previous tests showed that the template engine is not the bottleneck in the code generation. It is the optimizer. So compilation times would probably stay the same.

Currently, I am leaning to 2). Any thoughts?

**Comment by jcanizales** [comment]
As this project is aimed at showcasing what's possible in terms of unifying/generalizing C++ concepts and offering a simple UX to them, I strongly prefer your option 2.
